From: len.brown@intel.com
Subject: x86 intel: Set perf BIAS MSR to default value on boot
References: bnc#702604
Patch-Mainline: never, should be done in userspace at some point of time

https://lkml.org/lkml/2011/4/1/20

Some BIOSes may not intitialize perf BIAS MSR (a value from 0-15)
0 -> performance
15-> powersave
correctly.

Signed-off-by: Thomas Renninger <trenn@suse.de>

---
 arch/x86/include/asm/msr-index.h |    3 +++
 arch/x86/kernel/cpu/intel.c      |   14 ++++++++++++++
 2 files changed, 17 insertions(+)

Index: linux-3.0-tmp-jikos/arch/x86/include/asm/msr-index.h
===================================================================
--- linux-3.0-tmp-jikos.orig/arch/x86/include/asm/msr-index.h
+++ linux-3.0-tmp-jikos/arch/x86/include/asm/msr-index.h
@@ -259,6 +259,9 @@
 #define MSR_IA32_TEMPERATURE_TARGET	0x000001a2
 
 #define MSR_IA32_ENERGY_PERF_BIAS	0x000001b0
+#define ENERGY_PERF_BIAS_PERFORMANCE	0
+#define ENERGY_PERF_BIAS_NORMAL		6
+#define ENERGY_PERF_BIAS_POWERSWAVE	15
 
 #define MSR_IA32_PACKAGE_THERM_STATUS		0x000001b1
 
Index: linux-3.0-tmp-jikos/arch/x86/kernel/cpu/intel.c
===================================================================
--- linux-3.0-tmp-jikos.orig/arch/x86/kernel/cpu/intel.c
+++ linux-3.0-tmp-jikos/arch/x86/kernel/cpu/intel.c
@@ -456,6 +456,20 @@ static void __cpuinit init_intel(struct
 
 	if (cpu_has(c, X86_FEATURE_VMX))
 		detect_vmx_virtcap(c);
+
+	/*
+	 * Initialize MSR_IA32_ENERGY_PERF_BIAS if BIOS did not.
+	 * x86_energy_perf_policy(8) is available to change it at run-time
+	 */
+	if (cpu_has(c, X86_FEATURE_EPB)) {
+		u64 epb;
+
+		rdmsrl(MSR_IA32_ENERGY_PERF_BIAS, epb);
+		if ((epb & 0xF) == 0) {
+			epb = (epb & ~0xF) | ENERGY_PERF_BIAS_NORMAL;
+			wrmsrl(MSR_IA32_ENERGY_PERF_BIAS, epb);
+		}
+	}
 }
 
 #ifdef CONFIG_X86_32
